pkg/nodeagent/controller/healthcheck/kubelet.go

```diff
@@ -0,0 +1,264 @@
+}
+
+// HasLastInternalIP returns true if the node.InternalIP was stored.
+// Exported for testing.
+func (k *KubeletHealthChecker) HasLastInternalIP() bool {
+	return k.lastInternalIP.IsValid()
+}
+
+// SetKubeletHealthEndpoint set the kubeletHealthEndpoint.
+// Exported for testing.
+func (k *KubeletHealthChecker) SetKubeletHealthEndpoint(kubeletHealthEndpoint string) {
+	k.kubeletHealthEndpoint = kubeletHealthEndpoint
+}
+
+// Check performs the actual health check for the kubelet.
+func (k *KubeletHealthChecker) Check(ctx context.Context, node *corev1.Node) error {
+	log := logf.FromContext(ctx).WithName("kubelet")
+
+	// This mimics the old behavior of the kubelet-health-monitor which checks for
+	// to many NotReady->Ready toggles in a certain time and reboots the node in such a case.
+	if isNodeReady(node) && !k.NodeReady {
+		needsReboot := k.ToggleKubeletState()
+		log.Info("Kubelet became Ready", "readinessChanges", len(k.KubeletReadinessToggles), "timespan", toggleTimeSpan)
+		if needsReboot {
+			log.Info("Kubelet toggled between NotReady and Ready too often. Rebooting the node now")
+			k.recorder.Eventf(node, corev1.EventTypeWarning, "kubelet", "Kubelet toggled between NotReady and Ready at least %d times in a %s time window. Rebooting the node now", maxToggles, toggleTimeSpan)
+			err := k.dbus.Reboot()
+			if err != nil {
+				k.RevertToggleKubeletState()
+				k.recorder.Event(node, corev1.EventTypeWarning, "kubelet", "Rebooting the node failed")
+				return fmt.Errorf("rebooting the node failed %w", err)
+			}
+		}
+	}
+	k.NodeReady = isNodeReady(node)
+
+	err := k.ensureNodeInternalIP(ctx, node)
+	if err != nil {
+		return err
+	}
+
+	httpClient := &http.Client{
+		Timeout: 10 * time.Second,
+	}
+	request, err := http.NewRequestWithContext(ctx, http.MethodGet, k.kubeletHealthEndpoint, nil)
+	if err != nil {
+		log.Error(err, "Creating request to kubelet health endpoint failed")
+		return err
+	}
+	response, err := httpClient.Do(request)
+	if err != nil {
+		log.Error(err, "HTTP request to kubelet health endpoint failed")
+	}
+	if err == nil && response.StatusCode == http.StatusOK {
+		if k.firstFailure != nil {
+			log.Info("Kubelet is healthy again", "statusCode", response.StatusCode)
+			k.recorder.Event(node, corev1.EventTypeNormal, "kubelet", "Kubelet is healthy")
+			k.firstFailure = nil
+		}
+		return nil
+	}
+	if k.firstFailure == nil {
+		now := k.Clock.Now()
+		k.firstFailure = &now
+		log.Error(err, "Kubelet is unhealthy")
+		k.recorder.Event(node, corev1.EventTypeWarning, "kubelet", "Kubelet is unhealthy")
+	}
+
+	if time.Since(*k.firstFailure).Abs() < maxFailureDuration {
+		return nil
+	}
+
+	log.Error(err, "Kubelet is not healthy, restarting")
+	err = k.dbus.Restart(ctx, k.recorder, node, kubeletServiceName)
+	if err == nil {
+		k.firstFailure = nil
+	}
+	return err
+}
+
+// ensureNodeInternalIP mimics the old weird logic which restores the internalIP of the node
+// if this was lost for some reason.
```

Comment:
Should we not rather describe what our intend is here instead of referring to the "old behavior"? Soon, the old behavior will be deleted. Then the comment might be confusing.